### ISSUE TITLE
-Wunsafe-buffer-usage warns for range-based for loop

### ISSUE THREAD

**Issue by davidstone** [llvm/llvm-project#59665]
The following statically safe code

```c++
void f() {
	int a[1] = {};
	for (auto x : a) {
	}
}
```

causes `clang -Wunsafe-buffer-usage` to emit

```console
<source>:3:14: warning: unchecked operation on raw buffer in expression [-Wunsafe-buffer-usage]
        for (auto x : a) {
                    ^
1 warning generated.
Compiler returned: 0
```

See it live: https://godbolt.org/z/WdzKTzT9P

Passing an array as the argument to the range-based for loop is guaranteed to be a safe construct (at least from the "accessing the elements of the array" perspective, which is what this warning is trying to address), so this warning should not trigger for that usage.

Additionally, the error message isn't especially helpful because it doesn't tell me what I should do instead, but from what I can tell this is a fallback message where we don't have a great automated suggestion. However, that's not especially relevant in this case because it shouldn't warn at all.

@haoNoQ @ziqingluo-90

**Comment by haoNoQ** [comment]
Please note that this warning is still under active development. We're going to both improve automatic fix coverage and suppress false alarms of this kind in the upcoming months :)

In the general case we will not eliminate all false positives of this kind. When the code is complicated enough, the warning assumes that safe collections are still the better choice. We'll probably draw the line at whether proving that the code is safe requires data flow analysis or symbolic reasoning.

In this case, of course, flow sensitivity is not required; the range-based for loop is "syntactically" safe, so I think it's likely that we'll suppress it. Of course, using `std::array` still makes a lot of sense here.

In a manual for-loop case, if the loop is written in a very specific simple manner and none of the loop condition variables are modified in the body of the loop, it might still be possible to suppress the warning as well with purely syntactic analysis:
```c
void f() {
	int a[1] = {};
	for (size_t i = 0; i < 1; ++i) {
		int x = a[i];
	}
}
```
But anything extra on top of that would probably warrant a warning forever.

**Comment by davidstone** [comment]
Looks like this got fixed at some point
